common/src/device/transport/mqtt/ClientWebSocketChannel.cs

```diff
@@ -95,29 +95,35 @@ protected override void DoDisconnect()
             throw new NotSupportedException("ClientWebSocketChannel does not support DoDisconnect()");
         }
 
-        protected override async void DoClose()
+        protected override void DoClose()
         {
-            try
+            WebSocketState webSocketState = _webSocket.State;
+            if (webSocketState != WebSocketState.Closed && webSocketState != WebSocketState.Aborted)
             {
-                WebSocketState webSocketState = _webSocket.State;
-                if (webSocketState != WebSocketState.Closed && webSocketState != WebSocketState.Aborted)
-                {
-                    // Cancel any pending write
-                    CancelPendingWrite();
-                    _active = false;
+                // Cancel any pending write
+                CancelPendingWrite();
+                _active = false;
 
-                    using (var cancellationTokenSource = new CancellationTokenSource(TimeSpan.FromSeconds(30)))
+                Task.Run(async () =>
```

Comment:
Christian - we observed hangs in mqtt client and the recommendation from Max was to add a 5s delay before closing. Could you recommend another way where the client can close but we don't block it for 5s?

---
In reply to: [217779370](https://github.com/Azure/azure-iot-sdk-csharp/pull/619#discussion_r217779370) [](ancestors = 217779370)

Reply:
@CIPop to give some context here: I don't know about 5 sec but the dispatch there is to avoid blocking executor thread. WS client tends to deadlock on simultaneous closure. Locking CloseAsync while running on executor thread will not buy us anything. That said, I don't see why we couldn't await on that task except that if CloseAsync deadlocks we would still want to make progress wherever we await it.

Reply:
@alextolp, are you using .NET Framework or .NET Core when you're seeing the deadlock?

@nayato, according to MSDN Task.Run uses the default task scheduler (I read that as the _current_ task scheduler and might be wrong). Wouldn't that if written like this, it would still run in the dotNetty executor loop?